components/server/src/orgs/organization-service.ts

```diff
@@ -183,90 +184,103 @@ export class OrganizationService {
         orgId: string,
         memberId: string,
         role: OrgMemberRole,
+        txCtx?: TransactionalContext,
     ): Promise<void> {
         if (userId) {
             await this.auth.checkPermissionOnOrganization(userId, "write_members", orgId);
         }
-        if (role !== "owner") {
-            const members = await this.teamDB.findMembersByTeam(orgId);
-            if (!members.some((m) => m.userId !== memberId && m.role === "owner")) {
-                throw new ApplicationError(ErrorCodes.CONFLICT, "Cannot remove the last owner of an organization.");
-            }
-        }
-        const members = await this.teamDB.findMembersByTeam(orgId);
-        const firstMember = members.filter((m) => m.userId !== BUILTIN_INSTLLATION_ADMIN_USER_ID).length === 0;
-        if (firstMember) {
-            // first member (that is not an admin) is going to be an owner
-            role = "owner";
-            log.info({ userId: memberId }, "First member of organization, setting role to owner.");
-        }
-
+        let members: OrgMemberInfo[] = [];
         try {
-            await this.teamDB.transaction(async (db) => {
-                await db.addMemberToTeam(memberId, orgId);
-                await db.setTeamMemberRole(memberId, orgId, role);
+            await this.teamDB.transaction(txCtx, async (teamDB, txCtx) => {
+                members = await teamDB.findMembersByTeam(orgId);
+                const hasOtherRegularOwners =
+                    members.filter(
+                        (m) =>
+                            m.userId !== BUILTIN_INSTLLATION_ADMIN_USER_ID && //
+                            m.userId !== memberId && //
+                            m.role === "owner",
+                    ).length > 0;
+                if (!hasOtherRegularOwners) {
+                    // first regular member is going to be an owner
```

Comment:
The subtile change here is, that if the single owner would leave the org, the next sign-up would be promote to be an owner. before that, this role change would be only applied for the first regular member to join the org. 

Need to think about this case first. (preview env is spinning up...)

Reply:
the single owner cannot leave an org and cannot be downgraded to a member. The change is that before we would throw an exception when someone tried to downgrade the single owner, where now we change the role to owner no matter what the previous state was. I agree the logic seems too involved, i.e.e the method does too much. Let me think how we can simplify things. maybe having two methods makes more sense. addMember and setRole.